packages/pangea-sdk/pangea_django/pangea_django_auth.py

```diff
@@ -0,0 +1,122 @@
+# Copyright 2023 Pangea Cyber Corporation
+# Author: Pangea Cyber Corporation
+
+from datetime import datetime
+import os
+import traceback
+import string
+import secrets
+
+from django.contrib.auth.backends import BaseBackend
+from django.contrib.auth import get_user_model
+from django.contrib.auth.models import AnonymousUser
+
+from pangea.services.authn.authn import AuthN
+from pangea.config import PangeaConfig
+
+UserModel = get_user_model()
+
+def generate_state_param(request):
+    alphabet = string.ascii_letters + string.digits
+    state = ''.join(secrets.choice(alphabet) for i in range(12))
+    request.session["PANGEA_LOGIN_STATE"] = state
+    return state
+
+
+class PangeaAuthMiddleware():
+    class InvalidSessionException(Exception):
+        pass
+
+    def __init__(self, get_response):
+        self.authn = PangeaAuthentication()
+        self.get_response = get_response
+
+    def __call__(self, request):
+        active_token = None
+        try:
+            active_token = request.session.get("PANGEA_ACTIVE_TOKEN")
+            if active_token:
+                expires = active_token["expire"]
+                try:
+                    expires = datetime.fromisoformat(expires)
+                except ValueError:
+                    # support older python that doesnt understand the trailing Z
+                    expires = datetime.fromisoformat(expires.rstrip("Z"))
+                if expires < datetime.utcnow():
+                    refresh_token = request.session["PANGEA_REFRESH_TOKEN"]["token"]
+                    response = self.authn.client.refresh(refresh_token=refresh_token)
+                    if not response.status == 'Success':
+                        raise self.InvalidSessionException
+                    active_token = response["active_token"]
+                    response.session["PANGEA_ACTIVE_TOKEN"] = active_token
+            else:
+                raise self.InvalidSessionException
+        except self.InvalidSessionException as e:
+            request.user = AnonymousUser()
+        else:
+            request.user = self.authn.get_user(username=active_token["email"])
+
+        return self.get_response(request)
+
+class PangeaAuthentication(BaseBackend):
+    def __init__(self):
+        token = os.getenv("PANGEA_AUTHN_TOKEN")
+        domain = os.getenv("PANGEA_DOMAIN")
+        self.config = PangeaConfig(domain=domain)
+        self.authn = AuthN(token, config=self.config, logger_name="pangea")
+        super().__init__()
+
+    def authenticate(self, request):
+        code = request.GET.get('code')
+        state = request.GET.get('state')
+        print(request.GET)
+        expected_state = request.session.get("PANGEA_LOGIN_STATE")
```

Comment:
expected_state needs to be compared against state before proceeding.  after finding a match, PANGEA_LOGIN_STATE should be removed from the session object.

Reply:
this should fail if either state is not present in the request args or pangea_login_state is not present in the session store.